### ISSUE TITLE
Post-release flat storage refactor

### ISSUE THREAD

**Issue by Longarithm** [near/nearcore#8826]
### Goal

Improve flat storage quality for better engineering experience. Continues #8577 started in 2023'Q1.

### Priorities

* **P0**: should be implemented before MVP rollout. Involves addressing major pain points as well as changes to persistent data format (to avoid migration overhead in the near future).
* **P1**: important, but not a blocker for MVP. Can be implemented in parallel with MVP rollout.
* **P2**: nice-to-have improvements.

### Tasks

* [ ] Introduce a separate status for state sync + #8250
* [x] Address [review comments](https://github.com/near/nearcore/pull/8664#pullrequestreview-1326069272) in #8664
  - P1
* [ ] `FlatStorage` should be **the only API** to interact with flat storage for a fixed shard id.
  - P1
  - always create it and encapsulate `FlatStorageCreationStatus` (including `DontCreate` when `protocol_feature_flat_state` is disabled) as part of its state
  - no module outside `flat` should read/write to `FlatState*` column, instead it should use `FlatStorage` public API
  - move flat storage creation code from chain to `near_store::flat`
  - add `FlatStorage::create_for_genesis` to create flat storage instance for genesis (sets flat state head to genesis hash)
  - <del>get rid of `FlatStateFactory`, use `FlatStorage` instead</del> - decided to keep it as a container for `FlatStorage` instances across multiple shards.
* [ ] <del>chain should know very little (ideally nothing) about flat storage</del> - not really feasible/reasonable. This would result in even more bloated `RuntimeAdapter` interface and introduces a lot of indirections which would make the code harder to understand and navigate (see [this comment](https://github.com/near/nearcore/issues/8577#issuecomment-1433491047) for more context). Instead let's expose `FlatStorage` as a part of `RuntimeAdapter` trait so chain has call its methods directly.
  - P1
* [ ] remove `store_helper` module
  - P2
  - add storage-related methods to the corresponding structs (`FlatStorage`, `FlatStateDelta`, etc.)
* [ ] Uncomment and fix `test_state_sync` 
  - P2 as state sync is currently disabled
* [ ] Update `test_save_to_file` to also check for the flat storage data
  - P2
* [ ] If background migration is running but the block for which we create it got GC-d (see `Chain.clear_data`), we should interrupt node
  - P2 - nice-to-have for correctness
